### PR TITLE
Python Cleanup

### PR DESCRIPTION
**Pull request by donn**
+ Consolidated all existing DRC conversion functions into `drc_rosetta.py`
+ Consolidated fake_diode_replace and place_diode into `diode.py`
+ Consolidated and/or rewrote most def manipulation functions into `defutil.py` and other files under `odbpy/`.
+ Consolidated and/or rewrote most lef manipulation functions into `lefutil.py`
~ Moved all files using odb to a new folder, `odbpy/`
~ Move replicate/update config scripts to `scripts/config` folder
~ Move all in-line magic spells into proper files
~ Tcl formatting (4 space-indents on all modified files)
- Removed many unused files, all consolidated files
---
API Breaks

* Care was taken to ensure all OpenLane tcl commands have the same API presuming the API in question was not stupid.
    * Speaking of, `manual_macro_placement_test f` has been updated to `manual_macro_placement_test -f` to be consistent with literally every other command.
* All individual files that may be used by third party projects likely have a different CLI and are thus entirely broken.

---
[ci ets]

### CODE REVIEW

File: scripts/tcl_commands/all.tcl [hunk 1]
```diff
@@ -849,13 +876,14 @@ proc label_macro_pins {args} {
 
     set_if_unset arg_values(-pad_pin_name) ""
 
-    try_catch $::env(OPENROAD_BIN) -python $::env(SCRIPTS_DIR)/label_macro_pins.py\
+    try_catch $::env(OPENROAD_BIN) -python $::env(SCRIPTS_DIR)/odbpy/label_macro_pins.py\
         --lef $arg_values(-lef)\
-        --input-def $::env(CURRENT_DEF)\
+        --input-def \
```
Comment (kareefardi) [inline scripts/tcl_commands/all.tcl:881]: missing input def ?

File: scripts/tcl_commands/floorplan.tcl [hunk 2]
```diff
@@ -173,43 +173,43 @@ proc place_contextualized_io {args} {
...
 	parse_key_args "place_contextualized_io" args arg_values $options flags_map $flags
 
-	if {[file exists $arg_values(-def)] && [file exists $arg_values(-lef)]} {
-		TIMER::timer_start
-
-		file copy -force $arg_values(-def) $::env(placement_tmpfiles)/top_level.def
-		file copy -force $arg_values(-lef) $::env(placement_tmpfiles)/top_level.lef
+	if { ![file exists $arg_values(-def)] || ![file exists $arg_values(-lef)]} {
+		puts_err "Contextual IO placement: def/lef files don't exist, exiting"
+		return -code error
```
Comment (kareefardi) [inline scripts/tcl_commands/floorplan.tcl:178]: should this call flow_fail instead?

File: scripts/tcl_commands/magic.tcl [hunk 3]
```diff
@@ -164,38 +172,38 @@ proc run_magic_spice_export {args} {
...
-"
-if { \[info exist ::env(MAGIC_EXT_USE_GDS)\] && \$::env(MAGIC_EXT_USE_GDS) } {
-	gds read \$::env(CURRENT_GDS)
-} else {
-	lef read $::env(TECH_LEF)
-	if {  \[info exist ::env(EXTRA_LEFS)\] } {
-		set lefs_in \$::env(EXTRA_LEFS)
-		foreach lef_file \$lefs_in {
-			lef read \$lef_file
+	set commands "\
```
Comment (kareefardi) [inline scripts/tcl_commands/magic.tcl:175]: how do you suggest we get rid of this pattern ?

File: scripts/tcl_commands/placement.tcl [hunk 4]
```diff
@@ -101,15 +103,30 @@ proc add_macro_placement {args} {
...
+    if { [info exists flags_map(-fixed)] } {
+        lappend arg_list --fixed
     }
+
+    lappend arg_list $::env(CURRENT_DEF)
+    
+    try_catch $::env(OPENROAD_BIN) -python $::env(SCRIPTS_DIR)/odbpy/manual_macro_place.py {*}$arg_list |& tee $::env(TERMINAL_OUTPUT) [index_file $::env(placement_logs)/macro_placement.log]
+
+    TIMER::timer_stop
+    exec echo "[TIMER::get_runtime]" | python3 $::env(SCRIPTS_DIR)/write_runtime.py "detailed placement - openroad"
```
Comment (kareefardi) [inline scripts/tcl_commands/placement.tcl:129]: should this be called detailed placement ?

File: scripts/tcl_commands/routing.tcl [hunk 5]
```diff
@@ -236,17 +245,19 @@ proc power_routing {args} {
...
+    try_catch $::env(OPENROAD_BIN) -python $::env(SCRIPTS_DIR)/odbpy/power_utils.py\
+        power_route\
         --input-lef $arg_values(-lef)\
-        --input-def $arg_values(-def)\
         --core-vdd-pin $arg_values(-power)\
         --core-gnd-pin $arg_values(-ground)\
-        -o $arg_values(-output_def)\
-        {*}$arg_values(-extra_args) |& tee [index_file $::env(routing_logs)/power_routing.log] $::env(TERMINAL_OUTPUT)
+        --output $arg_values(-output_def)\
+        {*}$arg_values(-extra_args)
```
Comment (kareefardi) [inline scripts/tcl_commands/routing.tcl:254]: missing backslash

File: scripts/apply_def_template.py [hunk 6]
```diff
@@ -17,88 +17,89 @@
...
-
-    #read template Def
-    templateDEFOpener = open(templateDEF,"r")
-    if templateDEFOpener.mode == 'r':
-        templateDEFContent =templateDEFOpener.read()
+    subprocess.check_output(
+        [
+            "openroad",
+            "-python",
+            "move_pins" f"{scriptsDir}/odbpy/defutil.py",
```
Comment (kareefardi) [inline scripts/apply_def_template.py:60]: I am not sure but something looks odd here

File: .github/scripts/run_tests.py [hunk 7]
```diff
@@ -61,7 +72,10 @@
     if e.returncode != 2:
         raise e
 
-cat = lambda x: print(open(x).read())
+
+def cat(x):
+    print(open(x).read())
```
Comment (kareefardi) [inline .github/scripts/run_tests.py:77]: is it safe not to explicitly call close?

File: dependencies/tool.py [hunk 8]
```diff
@@ -55,33 +67,41 @@ def docker_args(self) -> List[str]:
...
-                in_install=tool['in_install'] if tool.get('in_install') is not None else True,
-                dependencies= tool.get('dependencies') or [],
-                pdk=tool.get('pdk') or False
+            final_dict[tool["name"]] = Tool(
+                name=tool["name"],
+                repo=tool["repo"],
+                commit=tool["commit"],
+                build_script=tool.get("build") or "",
+                default_branch=tool.get("default_branch") or None,
+                in_container=tool["in_container"]
```
Comment (kareefardi) [inline dependencies/tool.py:84]: this looks awkward but meh

File: scripts/tcl_commands/all.tcl (hunk 1 above)
Comment (donn) [reply scripts/tcl_commands/all.tcl]: Sorry, typo.

File: env.py [hunk 9]
```diff
@@ -100,99 +122,125 @@ def issue_survey():
...
             if not os.path.exists("/openlane"):
                 raise Exception()
             alert = "Alert: You appear to be running this script inside the OpenLane container. This is not the intended usecase for this file."
-            final_report += ("%s\n" % alert)
+            final_report += "%s\n" % alert
             print(alert, file=alerts)
-        except Exception as e:
-            alert = "Critical Alert: No Docker or Docker-compatible container engine was found."
-            final_report += ("%s\n" % alert)
+        except Exception:
```
Comment (kareefardi) [inline env.py:142]: missing as e ?

File: env.py [hunk 10]
```diff
@@ -100,99 +122,125 @@ def issue_survey():
...
+            alert = "Critical Alert: No Docker or Docker-compatible container engine was found: {e}."
```
Comment (kareefardi) [inline env.py:143]: also I think you didn't mean to put `{e}` as this is not and f string

File: .github/scripts/run_tests.py (hunk 7 above)
Comment (donn) [reply .github/scripts/run_tests.py:77]: yes

File: env.py (hunk 9 above)
Comment (donn) [reply env.py:142]: No, flake is not very merciful about unused variables.

File: env.py (hunk 10 above)
Comment (donn) [reply env.py:143]: yeah thanks

File: scripts/tcl_commands/magic.tcl (hunk 3 above)
Comment (donn) [reply scripts/tcl_commands/magic.tcl:175]: I have no immediate interest in doing so.

File: scripts/apply_def_template.py (hunk 6 above)
Comment (donn) [reply scripts/apply_def_template.py:60]: something IS odd here, there should be a comma

File: scripts/apply_def_template.py [hunk 11]
```diff
@@ -57,7 +57,8 @@ def remove_power_pins(DEF):
         [
             "openroad",
             "-python",
-            "move_pins" f"{scriptsDir}/odbpy/defutil.py",
+            "move_pins",
```
Comment (kareefardi) [inline scripts/apply_def_template.py:60]: shouldn't `move_pins` come after ` f"{scriptsDir}/odbpy/defutil.py"`
Also I think it is `replace_pins`

**Review by kareefardi — CHANGES_REQUESTED** [review]

File: scripts/apply_def_template.py (hunk 11 above)
Comment (donn) [reply scripts/apply_def_template.py]: Yes and yes. Oh god...

**Review by kareefardi — DISMISSED** [review]

**Comment by donn** [comment]
I'm gonna work longer on this. I might wanna add tests for all of this before merging it.

**Comment by donn** [comment]
A cursory review would be fine.

**Comment by antonblanchard** [comment]
Looks great to me @donn 

- Gets rid of the only use of perl
- Consolidates all the DRC conversion into a single file, together with examples of what each conversion does
- Moves the inline magic TCL script creation into separate files (something that always annoyed me)

**Review by antonblanchard — APPROVED** [review]

**Pull request merged**